nix/shell.go

```
@@ -0,0 +1,181 @@
+	// Launch a fallback shell if we couldn't find the path to the user's
+	// default shell.
+	if s.binPath == "" {
+		cmd := exec.Command("nix-shell", nixPath)
+		cmd.Stdin = os.Stdin
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+
+		debug.Log("Unrecognized user shell, falling back to: %v", cmd.Args)
+		return errors.WithStack(cmd.Run())
+	}
+
+	cmd := exec.Command("nix-shell", nixPath)
+	cmd.Args = append(cmd.Args, "--pure", "--command", s.execCommand())
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	debug.Log("Executing nix-shell command: %v", cmd.Args)
+	return errors.WithStack(cmd.Run())
+}
+
+// execCommand is a command that replaces the current shell with s.
+func (s *Shell) execCommand() string {
+	shellrc, err := writeDevboxShellrc(s.userShellrcPath, s.UserInitHook)
+	if err != nil {
+		debug.Log("Failed to write devbox shellrc: %v", err)
+		return "exec " + s.binPath
+	}
+
+	switch s.name {
+	case shBash:
+		return fmt.Sprintf(`exec /usr/bin/env ORIGINAL_PATH="%s" %s --rcfile "%s"`,
+			os.Getenv("PATH"), s.binPath, shellrc)
+	case shZsh:
+		return fmt.Sprintf(`exec /usr/bin/env ORIGINAL_PATH="%s" ZDOTDIR="%s" %s`,
+			os.Getenv("PATH"), filepath.Dir(shellrc), s.binPath)
+	case shKsh, shPosix:
+		return fmt.Sprintf(`exec /usr/bin/env ORIGINAL_PATH="%s" ENV="%s" %s `,
+			os.Getenv("PATH"), shellrc, s.binPath)
+	default:
+		return "exec " + s.binPath
+	}
+}
+
+func writeDevboxShellrc(userShellrcPath string, userHook string) (path string, err error) {
+	// We need a temp dir (as opposed to a temp file) because zsh uses
+	// ZDOTDIR to point to a new directory containing the .zshrc.
+	tmp, err := os.MkdirTemp("", "devbox")
+	if err != nil {
+		return "", fmt.Errorf("create temp dir for shell init file: %v", err)
+	}
+
+	// This is a best-effort to include the user's existing shellrc. If we
+	// can't read it, then just omit it from the devbox shellrc.
+	userShellrc, _ := os.ReadFile(userShellrcPath)
```

maybe only ignore error for `os.IsNotExist(err)`?
and set userShellrc to be empty-string in that case?

---

I think you're right here. It might be possible to return an error and a partially read file. I'll update it to explicitly set `userShellrc` to empty if there's an error so we don't risk running part of the user's shellrc.